pykafka/producer.py

```diff
@@ -231,14 +231,38 @@ def _setup_owned_brokers(self):
                     self, partition.leader)
         return queued_messages
 
-    def stop(self):
-        """Mark the producer as stopped"""
-        self._running = False
+    def _stop_owned_brokers(self):
         self._wait_all()
         if self._owned_brokers is not None:
             for owned_broker in self._owned_brokers.values():
                 owned_broker.stop()
 
+    def get_queue_readers(self):
+        if not self._owned_brokers:
+            return []
+        return [owned_broker._queue_reader_worker for owned_broker in self._owned_brokers.values() if owned_broker.running]
+
+    def stop(self, wait=True):
+        """Mark the producer as stopped, and wait until all messages to be sent
+
+        The join here works because new queue readers are spawned during the execution of the old ones i.e:
+        After a queue reader in its call to producer._send_request encounters either a SocketDisconnectedError or
+        NotLeaderForPartition.ERROR_CODE it updates the cluster, sets up owned brokers thus starting new queue reader
+        threads. Because of this when we call self.get_queue_readers, we get the newly created queue readers, and so try
+        to stop and join them etc ... until they all stop without encountering problems in producer._send_request
+
+        :param wait: whether we should wait until all messages to be sent or not
+        :type wait: bool
+        """
+        while wait and self._running:
+            queue_readers = self.get_queue_readers()
+            self._stop_owned_brokers()
+            if len(queue_readers) == 0:
+                self._running = False
+            else:
+                for queue_reader in queue_readers:
+                    queue_reader.join()
+
```

Comment:
If `wait` is `False`, we'll never set `self._running = False`, which seems wrong. Maybe you could put another `self._running = False` after the end of the loop.